src/Middleware/RequestThrottling/src/Policies/StackPolicy.cs

```diff
@@ -0,0 +1,76 @@
+using System;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.RequestThrottling.Policies
+{
+    class StackPolicy : IQueuePolicy
+    {
+        private TaskCompletionSource<bool>[] _buffer;
+        private int _head; 
+        private int _queueLength;
+
+        private Task<bool> _trueTask = Task.FromResult(true);
+
+        private object _bufferLock = new Object();
+
+        private int _freeServerSpots;
+
+        public StackPolicy(IOptions<TailDropOptions> options)
+        {
+            _buffer = new TaskCompletionSource<bool>[options.Value.RequestQueueLimit];
```

Comment:
Should _buffer be a `List<T>` instead of an array? An array will immediately allocate up to the limit, and if someone specifies `TailDropOptions.RequestQueueLimit = int.MaxValue` then they're going to get an OutOfMemoryException.

Reply:
My thought process was that I'm trying to optimize for the worst-case scenario. When the server is under maximum load, the entire buffer will be allocated anyways, so I would rather error early than during a traffic spike.

Reply:
I agree that it makes sense to lazily grow the collection, and List is probably the best way to do that.

The List should be a lot better in cases with a large RequestQueueLimit but not many queued requests in practice. Even in the worst case when the queue fills up, a List shouldn't be that much less efficient than an array. After all, List is [backed by a single array](https://github.com/dotnet/corefx/blob/04e05d7d0a02f43ce5e6df455052dbfc0862dff7/src/Common/src/CoreLib/System/Collections/Generic/List.cs#L115) anyway. The array just gets reallocated [with double the previous capacity](https://github.com/dotnet/corefx/blob/04e05d7d0a02f43ce5e6df455052dbfc0862dff7/src/Common/src/CoreLib/System/Collections/Generic/List.cs#L410) and copied as the list grows. So there should be at most log(N) reallocations/copies where N is the RequestQueueLimit, and it should be as fast as an array once the queue stops growing.

Reply:
There is now a list.